Reworks parameters again, fixes issue #288

jamonholmgren opened the pull request:
This removes a lot of the janky manual parsing of command line arguments that we were doing before and uses `minimist` for all of it.

My confidence level that this works across all situations is not super high, since I did rewrite a fair amount of critical path code here. Tests all pass, and in fact some jankiness with the tests is now looking much more like I'd expect, so that's a good sign.

jamonholmgren commented on src/domain/runtime.js:
```diff
@@ -60,32 +60,30 @@ async function run (rawCommand, options) {
     commandArray = commandArray.slice(2)
   }
 
-  // find the pluginName and commandName from the cli args
```
Removed this whole block of code, using minimist via the `parseParams` function now.

jamonholmgren commented on src/domain/runtime.js:
```diff
@@ -94,15 +92,6 @@ async function run (rawCommand, options) {
     (this.defaults && this.defaults[context.plugin.name]) || {}
   )
 
-  // normalized parameters
```
Not necessary, using the new `createParams` function above.

jamonholmgren commented on src/domain/runtime-parameters.test.js:
```diff
@@ -4,7 +4,7 @@ const Runtime = require('./runtime')
 test('can pass arguments', async t => {
   const r = new Runtime()
   r.load(`${__dirname}/../fixtures/good-plugins/args`)
-  const { command, parameters } = await r.run('args hello steve kellock', { caps: false })
```
I never really liked how we had to put the plugin name here. It's no longer necessary.

I'm not _completely_ sure this is the right direction, though. I'm having some trouble wrapping my mind around the plugins command invocation and what @skellock intended here.

It does feel right, though.

skellock replied:
The plugin doesn't make much sense in the new world order.  It's commands all the way down.

jamonholmgren replied:
I’m so glad you said that. It makes my life a lot easier. Ha

jamonholmgren commented on src/domain/runtime-parameters.test.js:
```diff
@@ -32,3 +32,17 @@ test('can pass arguments, even with nested alias', async t => {
+test('can pass arguments with mixed options', async t => {
+  const r = new Runtime()
+  r.load(`${__dirname}/../fixtures/good-plugins/args`)
+  const { command, parameters } = await r.run('--chocolate=true --foo -n 1 hello steve kellock')
```
New regression test for #288 .

skellock commented on src/utils/cli/normalize-params.js:
```diff
@@ -1,20 +1,28 @@
+function parseParams (argv, extraOpts = {}) {
+  // chop it up minimist!
+  const parsed = minimist(argv)
+  const array = parsed._.slice()
```
`._.` <--- awww... its so cute

jamonholmgren replied:
Haha didn’t even notice that.

skellock replied:
can't unsee

skellock commented:
One small thing i've noticed.

`am --force generate`

But these do:

```sh
am generate
am generate models
am generate --force
am generate models --force
am generate --force models
```

I don't think it's a deal breaker.  It'd be nice to support this though.  I wouldn't hold up this PR because of it though.

jamonholmgren commented:
@skellock So this is a minimist issue, and essentially from what I can tell it's a design decision.

![image](https://user-images.githubusercontent.com/1479215/34008559-966dbecc-e0ba-11e7-91ac-bc3f63b14a4c.png)

See the problem here? It's using the next arg as input. So it'll set `g` to the next argument. minimist wouldn't work with npm either. 😞 🐼 '

Given [this issue](https://github.com/substack/minimist/issues/115), and given that we _already had_ `yargs-parser` as a dependency (through a series of unfortunate events starting with `nyc`), I switched over to `yargs-parser` and removed `minimist`. yargs-parser exhibits the same exact behavior as above, so I think we will lean on the CLI creator to preprocess situations like `npm -g install` before handing it off to gluegun's runtime. Sound reasonable?

jamonholmgren commented:
Semaphore is having issues, but tests are passing just fine.

skellock commented:
`yargs-parser` wouldn't have been a production dependency tho because `nyc` is a dev dependency.

But this is all good with me.  `minimist` was internal anyway, so good call.

jamonholmgren commented:
Oh, you're right. Didn't think about devDependencies. 👍

infinitered-circleci commented:
:tada: This PR is included in version 0.22.1 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/gluegun)
- [GitHub release](https://github.com/infinitered/gluegun/releases/tag/v0.22.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
